crates/bevy_reflect/src/reflect.rs

```diff
@@ -46,17 +49,15 @@ pub enum ReflectMut<'a> {
 ///
 /// When using `#[derive(Reflect)]` with a struct or tuple struct, the suitable subtrait for that
 /// type (`Struct` or `TupleStruct`) is derived automatically.
-///
-/// # Safety
-/// Implementors _must_ ensure that [`Reflect::any`] and [`Reflect::any_mut`] both return the `self`
-/// value passed in. If this is not done, [`Reflect::downcast`](trait.Reflect.html#method.downcast)
-/// will be UB (and also just logically broken).
-pub unsafe trait Reflect: Any + Send + Sync {
-    /// Returns the [type name] of the underlying type.
-    ///
-    /// [type name]: std::any::type_name
+pub trait Reflect: Any + Send + Sync {
+    /// Returns the [type name][std::any::type_name] of the underlying type.
     fn type_name(&self) -> &str;
 
+    /// Returns the [`TypeId`] of the underlying type.
+    fn underlying_type_id(&self) -> TypeId {
+        TypeId::of::<Self>()
+    }
```

Comment:
Thinking more on how this might be confusing for users— especially those who are trying to create their own proxy containers— I'm wondering if including this method is justified.

My thinking is this:
* Type name indicates the represented type. It returns what the data should be treated like and not the actual underlying type. For example, `dyn_struct.type_name()` might return `my_crate::Foo` instead of `bevy_reflect::DynamicStruct`. In my opinion, that's _okay_ since it's useful for verifying a proxy is meant to represent a specific data structure.
* `TypeId`, on the other hand, indicates the actual type. And this can be useful to then verify that the type a value represents is actually that type.

Now those beliefs are completely arbitrary haha (it could honestly go either way or no way at all), but I think it's useful to make some sort of distinction like that. Plus, this could cause issues since `is` uses `underlying_type_id` under the hood, which could return false positives for custom proxy containers if the implementor isn't careful or doesn't understand.

---

All that to say, I think we should either just use [`Any::type_id`](https://doc.rust-lang.org/std/any/trait.Any.html#tymethod.type_id) directly (and remove this method) or add more documentation on this method stating what "underlying" means and what a user can expect to be returned.

Reply:
Right now this method is pretty much just placeholder; originally I included it to allow more explicit control over how `is` behaves, but in light of the Unique Reflect RFC I'm uncertain whether this should be included at all. It certainly isn't good how it is but I can't begin to decide how it should be until Unique Reflect is finished/implemented.

Reply:
`Any::type_id` is now used directly